Add a 2nd time derivative auxkernel

GiudGiud opened the pull request:
I wanted to check on error messages for using 2nd order time derivatives with time integration schemes that dont implement them. It turns out it's safe, the time integrator does not request the 2nd time derivative and this brings an error.

but then I was mostly ready for this PR so just made it. 

refs #23770

GiudGiud committed:
Add second order time derivative auxkernel using Coupleable
 as dotdot is not in functors yet (and no real need for now)

GiudGiud committed:
Add test for SecondTimeDerivativeAux, refs #23770

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1410365/) on 67de622 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23771/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1410401/) on 67de622 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ae61f4</th>
      <th colspan=4>#23771 67de62</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.83%</td>
      <td>84.84%</td>
      <td>+0.01%</td>
      <td>89.66%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>84017</td>
      <td>84050</td>
      <td>+33</td>
      <td>26</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15026</td>
      <td>15022</td>
      <td>-4</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23771/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23771/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23771/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 89.66% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented on framework/include/auxkernels/SecondTimeDerivativeAux.h:
```diff
@@ -0,0 +1,37 @@
+protected:
+  virtual Real computeValue() override;
+
+  /// Variable
```
```suggestion
  /// Variable we're computing the second time derivative of
```

lindsayad commented on framework/src/auxkernels/SecondTimeDerivativeAux.C:
```diff
@@ -0,0 +1,62 @@
+Real
+SecondTimeDerivativeAux::computeValue()
+{
+  if (_use_qp_arg)
+  {
+    const auto qp_arg = std::make_tuple(_current_elem, _qp, _qrule);
+    return _factor(qp_arg) * _v[_qp];
+  }
+  else
+  {
+    const auto elem_arg = makeElemArg(_current_elem);
+    return _factor(elem_arg) * _v[_qp];
```
I don't really like the mixing of the on-the-fly and pre-init'd data paradigms. It suggests we're missing capability one way or the other. Do you need one capability from one and one from another?

GiudGiud replied:
we are missing the dotdot in the functors. 
But is it really worth adding before we have a use for them?

lindsayad replied:
I don't think it's worth it ... so my suggestion would be to not use functors for `_factor` ... but I suppose that `TimeDerivativeAux` uses functors for the factor and if `_factor` is really something people will use then it's annoying to support multiple parameter options, e.g. Function/CoupledVar/MaterialProperty ... so this is probably fine

lindsayad commented on test/tests/auxkernels/time_derivative_second_aux/test.i:
```diff
@@ -0,0 +1,87 @@
+
+[AuxKernels]
+  # Time derivative of a nonlinear variable
+  # Time derivative of an aux variable is not currently supported
+  # in the Auxiliary system (only in nonlinear)
```
really? I believe you, but I know we attach time integrators to the auxiliary system so conceptually I feel like we should be able to. Do you get an error?

GiudGiud replied:
this is pasted from the other test, where I wrote this most likely

I ll see what I can find

GiudGiud replied:
ok seems to work. I tried two examples and got something reasonable with two integration schemes. 

I dont recall why I added this note

lindsayad commented on test/tests/auxkernels/time_derivative_second_aux/tests:
```diff
@@ -0,0 +1,30 @@
+    input = 'test.i'
+    exodiff = 'test_out.e'
+    cli_args = "AuxVariables/inactive='variable_derivative_fv' AuxKernels/inactive='var_derivative_to_fv'"
+    requirement = "The system shall be able to compute the second time derivative of a non linear variable."
```
```suggestion
    requirement = "The system shall be able to compute the second time derivative of a nonlinear variable."
```

lindsayad commented on test/tests/auxkernels/time_derivative_second_aux/tests:
```diff
@@ -0,0 +1,30 @@
+    exodiff = 'test_out.e'
+    cli_args = "AuxVariables/inactive='variable_derivative_fv' AuxKernels/inactive='var_derivative_to_fv'"
+    requirement = "The system shall be able to compute the second time derivative of a non linear variable."
+    ad_indexing_type = 'global'
```
Ugh you have so many PRs that conflict with #23665 haha

lindsayad commented on test/tests/auxkernels/time_derivative_second_aux/tests:
```diff
@@ -0,0 +1,30 @@
+      input = 'test.i'
+      cli_args = 'AuxVariables/nodal/family=LAGRANGE AuxKernels/var_derivative/variable=nodal'
+      expect_err = 'This AuxKernel only supports Elemental fields'
+      detail = 'a second time derivative is requested but the variable is a nodal variable which is currently unsupported'
```
```suggestion
      detail = 'a second time derivative is requested but the variable is a nodal variable which is currently unsupported, and'
```

lindsayad commented on test/tests/auxkernels/time_derivative_second_aux/tests:
```diff
@@ -0,0 +1,30 @@
+      input = 'test.i'
+      cli_args = "AuxVariables/inactive='' AuxKernels/inactive=''"
+      expect_err = "'v' argument is a finite element variable but 'variable' is not"
+      detail = 'a second time derivative for a finite element variable is to be saved in a finite volume variable'
```
```suggestion
      detail = 'a second time derivative for a finite element variable is requested to be stored in a finite volume variable.'
```

GiudGiud committed:
Apply suggestions from code review

- tests file formatting
- improved docstrings
- remove note about time derivative of aux variables

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

lindsayad reviewed (changes requested):
I'll submit a patch for `ad_indexing_type`

lindsayad committed:
Remove ad_indexing_type

GiudGiud commented:
Interesting that both u and the derivative differ. 
Seems like tightening u might be all that's needed

GiudGiud commented:
fix is merged.

so the issue here is that the CD time integrator only supports the LINEAR solve_type, which has a very loose base tolerance (1e-5)
